File: jieun/5_징검다리.py

def solution(stones, k):
    N = len(stones)
    right = list(range(N+2)) # 왼쪽, 오른쪽 끝 패딩
    left = list(range(N+2))

    def find_right(x):
        if right[x] != x:
            right[x] = find_right(right[x])
        return right[x]
    
    def find_left(x):
        if left[x] != x:
            left[x] = find_left(left[x])
        return left[x]

    def union_right(x, y):
        x = find_right(x)
        y = find_right(y)
        if x > y:
            right[y] = x
        else:
            right[x] = y
            
    def union_left(x, y):
        x = find_left(x)
        y = find_left(y)
        if x < y:
            left[y] = x
        else:
            left[x] = y
    
    q = [(val, idx+1) for idx, val in enumerate(stones)]
    q.sort() # 디딤돌 값 오름차순 정렬
    # print("q:", q)
    last_cnt = 0
    
    for cur_val, cur_idx in q:
        last_cnt = max(last_cnt, cur_val)

        union_left(cur_idx, cur_idx-1)
        union_right(cur_idx, cur_idx+1)

        left_group = find_left(cur_idx)
        right_group = find_right(cur_idx)
        
        # print(f"(val:{cur_val}, idx:{cur_idx}) left_group: {left_group}, right_group:{right_group}")
        
        if right_group - left_group > k:
            break
    
    # last_cnt 번째 친구가 건넌 후 더 이상 징검다리 못 건넌다.
    answer = last_cnt 

    return answer


# stones = [1, 5]
# k = 2

stones = [2, 4, 5, 3, 2, 1, 4, 2, 5, 1]
k = 3

print("sol:", solution(stones, k))

File: jieun/5_징검다리.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

#define MAX 200000

vector<int> rightg(MAX + 2);
vector<int> leftg(MAX + 2);

int find_left(int x)
{
    if (leftg[x] == x)
        return x;
    return leftg[x] = find_left(leftg[x]);
}

int find_right(int x)
{
    if (rightg[x] == x)
        return x;
    return rightg[x] = find_right(rightg[x]);
}

void union_left(int x, int y)
{
    x = find_left(x);
    y = find_left(y);
    if (x < y)
        leftg[y] = x;
    else
        leftg[x] = y;
}

void union_right(int x, int y)
{
    x = find_right(x);
    y = find_right(y);
    if (x > y)
        rightg[y] = x;
    else
        rightg[x] = y;
}

int solution(vector<int> stones, int k)
{
    int answer = 0;
    int N = stones.size();

    vector<pair<int, int>> values(N);
    for (int i = 0; i < N; i++)
    {
        values[i] = make_pair(stones[i], i + 1);
    }
    sort(values.begin(), values.end());

    for (int i = 0; i < N + 2; i++)
    {
        rightg[i] = i;
        leftg[i] = i;
    }

    int last_cnt = 0;
    for (int i = 0; i < values.size(); i++)
    {
        int cur_val = values[i].first;
        int cur_idx = values[i].second;
        last_cnt = max(last_cnt, cur_val);
        union_left(cur_idx, cur_idx - 1);
        union_right(cur_idx, cur_idx + 1);
        if (find_right(cur_idx) - find_left(cur_idx) > k)
            break;
    }

    answer = last_cnt;

    return answer;
}